File: Algorithms/C++/Easy/252_Meeting_Rooms.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
bool IntervalSorter(Interval const& lhs, Interval const& rhs) {
    return lhs.start < rhs.start;
}

class Solution {
public:
    bool canAttendMeetings(vector<Interval>& intervals) {
        if (intervals.size() <= 1) {
            return true;
        }

        sort(intervals.begin(), intervals.end(), &IntervalSorter);

        for (int i = 0; i < intervals.size() - 1; i++) {
            if (intervals[i].end > intervals[i + 1].start) {
                return false;
            }
        }

        return true;
    }
};

File: Algorithms/Swift/Easy/252_Meeting_Rooms.swift
/**
 * Definition for an interval.
 * public class Interval {
 *   public var start: Int
 *   public var end: Int
 *   public init(_ start: Int, _ end: Int) {
 *     self.start = start
 *     self.end = end
 *   }
 * }
 */
class Solution {
    func canAttendMeetings(intervals: [Interval]) -> Bool {
        if intervals.isEmpty {
            return true
        }

        let sortedIntervals = intervals.sort { $0.start < $1.start }

        for i in 0..<sortedIntervals.count - 1 {
            let before = sortedIntervals[i]
            let after = sortedIntervals[i + 1]

            if before.end > after.start {
                return false
            }
        }

        return true
    }
}
